### ISSUE TITLE
DnsResolverProvider fails on Android

### ISSUE THREAD

**Issue by Erhannis** [eclipse/vert.x#2110]
Note: I originally posted this in https://github.com/vert-x3/vertx-proton/issues/54, but was told that issues with DnsResolverProvider should go here.  Note further that I'm using io.vertx:vertx-proton:3.4.2.

----

The following code (extracted from my test app) 

        File cache = new File(ctx.getCacheDir(), ".vertx");
        cache.mkdirs();
        System.setProperty(CACHE_DIR_BASE_PROP_NAME, cache.getAbsolutePath());

        //System.setProperty("vertx.disableDnsResolver", "true");

        Vertx vertx = Vertx.vertx();
        Context vCtx = vertx.getOrCreateContext();

        vCtx.runOnContext(x -> {
            ProtonClient client = ProtonClient.create(vertx);

            // Connect, then use the event loop thread to process things thereafter
            client.connect(new ProtonClientOptions().setSsl(true),"images.google.com", 5671, connectResult -> {
                 System.out.println("Inside");
            });
        });

never reaches the println.  (This should perhaps be a separate issue, but note that CACHE_DIR_BASE_PROP_NAME needs to be set, on Android, or things fail, too.  Note further that I'm using "images.google.com" as an example - it produces the same errors as I see when I give the hostname of our company's test server.  Port 5671 is not open on images.google.com, but since the failure of the DNS resolution is the issue I'd like to point out, it doesn't matter.)
Instead, it yields the following error:

    08-31 16:43:13.351 9040-9493/com.example.android.amqptest W/DnsNameResolver: [id: 0xc6cfbf96, L:/:::49431] Unexpected exception: 
                                                                             io.netty.handler.codec.DecoderException: java.lang.IllegalArgumentException: Invalid input to toASCII: .
                                                                                 at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:98)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
                                                                                 at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
                                                                                 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
                                                                                 at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
                                                                                 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
                                                                                 at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
                                                                                 at java.lang.Thread.run(Thread.java:762)
                                                                              Caused by: java.lang.IllegalArgumentException: Invalid input to toASCII: .
                                                                                 at java.net.IDN.toASCII(IDN.java:112)
                                                                                 at java.net.IDN.toASCII(IDN.java:134)
                                                                                 at io.netty.handler.codec.dns.AbstractDnsRecord.<init>(AbstractDnsRecord.java:72)
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRawRecord.<init>(DefaultDnsRawRecord.java:61)
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:101)
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:71)
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:110)
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:68)
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33)
                                                                                 at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341) 
                                                                                 at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) 
                                                                                 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) 
                                                                                 at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479) 
                                                                                 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441) 
                                                                                 at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) 
                                                                                 at java.lang.Thread.run(Thread.java:762) 
                                                                              Caused by: Found zero length lable after NamePrep.. line:  0. preContext:  . postContext: 
                                                                             
                                                                                 at android.icu.impl.IDNA2003.convertToASCII(IDNA2003.java:185)
                                                                                 at android.icu.impl.IDNA2003.convertIDNToASCII(IDNA2003.java:275)
                                                                                 at android.icu.text.IDNA.convertIDNToASCII(IDNA.java:652)
                                                                                 at java.net.IDN.toASCII(IDN.java:110)
                                                                                 at java.net.IDN.toASCII(IDN.java:134) 
                                                                                 at io.netty.handler.codec.dns.AbstractDnsRecord.<init>(AbstractDnsRecord.java:72) 
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRawRecord.<init>(DefaultDnsRawRecord.java:61) 
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:101) 
                                                                                 at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:71) 
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:110) 
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:68) 
                                                                                 at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33) 
                                                                                 at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341) 
                                                                                 at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) 
                                                                                 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) 
                                                                                 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) 
                                                                                 at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565) 
                                                                                 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479) 
                                                                                 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441) 
                                                                                 at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) 
                                                                                 at java.lang.Thread.run(Thread.java:762) 

along with thousands of repetitions of

    08-31 13:45:18.988 21273-21301/com.example.android.amqptest W/NioEventLoop: Selector.select() returned prematurely 512 times in a row; rebuilding Selector sun.nio.ch.PollSelectorImpl@cc2ffa4.
    08-31 13:45:18.989 21273-21301/com.example.android.amqptest I/NioEventLoop: Migrated 2 channel(s) to the new Selector.

Something about the `ADDITIONAL` DNS section only having a period for the name, or something.

Now, when I comment out that line near the top of the code:

        System.setProperty("vertx.disableDnsResolver", "true");

things work fine.  (In the case of Google, the connection eventually times out, as expected.  On our test server, the connection succeeds, and everything works great.)  I therefore suspect that something is wrong with your DnsResolverProvider when used on Android.

**Comment by vietj** [comment]
can you try with this branch that updates to a newer Netty and perhaps fixes the issue https://github.com/eclipse/vert.x/tree/netty-4.1.15.Final ?

**Comment by Erhannis** [comment]
Well, once I got that branch to build...

(io.vertx.core.json.Json:170 errored that mapper.readValue was ambiguous.  I copied a change from commit e29fa50, casting the stream to DataInput.  Then some of the tests failed.  Skipping the tests, it build, and after rejiggering some version number I got vertx-proton to build with it.)

...then it did seem to work properly, with or without vertx.disableDnsResolver.
